src/wabt.cpp

```diff
@@ -0,0 +1,265 @@
+  wabt::interp::TableImport* import,
+  wabt::interp::Table* table,
+  const ErrorCallback& callback
+) {
+  (void)import;
+  (void)table;
+  (void)callback;
+  return wabt::Result::Error;
+}
+
+interp::Result WabtEthereumInterface::wabtUseGas(
+  const interp::HostFunc* func,
+  const interp::FuncSignature* sig,
+  Index num_args,
+  interp::TypedValue* args,
+  Index num_results,
+  interp::TypedValue* out_results,
+  void* user_data
+) {
+  (void)func;
+  (void)sig;
+  (void)num_results;
+  (void)out_results;
+
+  heraAssert(num_args == 1, "Invalid number of args");
+
+  WabtEthereumInterface *interface = reinterpret_cast<WabtEthereumInterface*>(user_data);
+
+  int64_t gas = static_cast<int64_t>(args[0].value.i64);
+
+  // FIXME: handle host trap here
+  interface->eeiUseGas(gas);
+
+  return interp::Result::Ok;
+}
+
+interp::Result WabtEthereumInterface::wabtFinish(
+  const interp::HostFunc* func,
+  const interp::FuncSignature* sig,
+  Index num_args,
+  interp::TypedValue* args,
+  Index num_results,
+  interp::TypedValue* out_results,
+  void* user_data
+) {
+  (void)func;
+  (void)sig;
+  (void)num_results;
+  (void)out_results;
+
+  heraAssert(num_args == 2, "Invalid number of args");
+
+  WabtEthereumInterface *interface = reinterpret_cast<WabtEthereumInterface*>(user_data);
+
+  uint32_t offset = args[0].value.i32;
+  uint32_t length = args[1].value.i32;
+
+  // FIXME: handle host trap here
+  interface->eeiRevertOrFinish(false, offset, length);
+
+  return interp::Result::Ok;
+}
+
+interp::Result WabtEthereumInterface::wabtGetCallDataSize(
+  const interp::HostFunc* func,
+  const interp::FuncSignature* sig,
+  Index num_args,
+  interp::TypedValue* args,
+  Index num_results,
+  interp::TypedValue* out_results,
+  void* user_data
+) {
+  (void)func;
+  (void)sig;
+  (void)num_results;
+  (void)out_results;
+  (void)args;
+  (void)num_args;
+
+  WabtEthereumInterface *interface = reinterpret_cast<WabtEthereumInterface*>(user_data);
+
+  out_results[0].type = sig->result_types[0];
+  out_results[0].value.i32 = interface->eeiGetCallDataSize();
+
+  return interp::Result::Ok;
+}
+
+ExecutionResult WabtEngine::execute(
+  evmc_context* context,
+  vector<uint8_t> const& code,
+  vector<uint8_t> const& state_code,
+  evmc_message const& msg,
+  bool meterInterfaceGas
+) {
+  (void)context;
+  (void)code;
+  (void)state_code;
+  (void)msg;
+  (void)meterInterfaceGas;
+
+  HERA_DEBUG << "Executing..." << endl;
+
+  // This is the wasm state
+  wabt::interp::Environment env;
+
+  // Lets instantiate our state
+  ExecutionResult result;
+  // FIXME: this only uses memory index 0 ...
+  WabtEthereumInterface* interface = new WabtEthereumInterface(context, state_code, msg, result, meterInterfaceGas, env.GetMemory(0));
+
+  // Lets add our host module
+  wabt::interp::HostModule* hostModule = env.AppendHostModule("ethereum");
+  heraAssert(hostModule, "Failed to create host module.");
+  hostModule->import_delegate.reset(interface);
```

Comment:
This requires a pointer and if I've used a `shared_ptr` it didn't work. Probably my fault.

Reply:
Move it down to:
```
  // FIXME: really bad design
  interface->setWasmMemory(env.GetMemory(0));
  hostModule->import_delegate = std::move(interface);
```

Reply:
The `import_delegate` must be set before the loading part unfortunately (before `ReadBinaryInterp`).